### PR TITLE
feat: add rdkafka2 plugin

### PR DESCRIPTION
**Pull request by onecer**
fix: https://github.com/kube-logging/logging-operator/issues/1373

### CODE REVIEW

File: v1.16-4.8/Dockerfile [hunk 1]
```diff
@@ -39,6 +40,7 @@ RUN apk add --no-cache \
          google-protobuf:3.21.12 \
          nio4r:2.7.3 \
          snappy:0.3.0 \
+         rdkafka:0.16.0 \
```
Comment (OverOrion) [inline v1.16-4.8/Dockerfile:43]: Thank you for your contribution @onecer, please add `rdkafka` to the `outputs/Gemfile` and update the `Gemfile.lock` as well.

File: v1.16/Dockerfile [hunk 2]
```diff
@@ -38,6 +39,7 @@ RUN apk add --no-cache \
          google-protobuf:3.21.12 \
          nio4r:2.6.0 \
          snappy:0.3.0 \
+         rdkafka:0.16.0 \
```
Comment (OverOrion) [inline v1.16/Dockerfile:42]: Thank you for your contribution @onecer, please add `rdkafka` to the `outputs/Gemfile` and update the `Gemfile.lock` as well.

File: v1.16-4.8/Dockerfile [hunk 3]
```diff
@@ -30,6 +30,7 @@ RUN apk add --no-cache \
  && apk add --no-cache $BUILD_DEPS \
  && apk add --no-cache "libssl3>=3.1.4-r1" \
  && apk add --no-cache "libcrypto3>=3.1.4-r1" \
+ && apk add --no-cache "bash" \
```
Comment (OverOrion) [inline v1.16-4.8/Dockerfile:33]: Why is bash required?

**Review by OverOrion — CHANGES_REQUESTED** [review]

File: v1.16-4.8/Dockerfile (hunk 1 above)
Comment (onecer) [reply v1.16-4.8/Dockerfile:43]: Okay

File: v1.16-4.8/Dockerfile (hunk 3 above)
Comment (onecer) [reply v1.16-4.8/Dockerfile:33]: > Why is bash required?

when I install rdkafka. It is required bash.

File: v1.16-4.8/Dockerfile (hunk 1 above)
Comment (onecer) [reply v1.16-4.8/Dockerfile:43]: > Thank you for your contribution @onecer, please add `rdkafka` to the `outputs/Gemfile` and update the `Gemfile.lock` as well.

We cannot do it that way. This plugin required `make`  to build it. It is a c extension. 

```
make DESTDIR\= sitearchdir\=./.gem.20240710-7-kjbl0d sitelibdir\=./.gem.20240710-7-kjbl0d
make failedNo such file or directory - make
```

In Dockerfile build stage, before we start install Gemfile.outputs. `make` have been deleted.

```
 && apk del $BUILD_DEPS \
```

File: v1.16/Dockerfile (hunk 2 above)
Comment (onecer) [reply v1.16/Dockerfile:42]: same as https://github.com/kube-logging/fluentd-images/pull/141#discussion_r1671790820

File: v1.16-4.8/Dockerfile [hunk 4]
```diff
@@ -9,7 +9,7 @@ ARG BUILD_DEPS=" \
       wget bzip2 zlib-dev git linux-headers \
       automake autoconf libtool build-base \
       ruby-dev libc6-compat geoip-dev \
-      snappy-dev gnupg \
+      snappy-dev gnupg bash \
```
Comment (onecer) [inline v1.16-4.8/Dockerfile:12]: add bash to $BUILD_DEPS, rdkafka plugin needs bash and build tools to build. after build done , $BUILD_DEPS will be remove.

File: v1.16-4.8/Dockerfile (hunk 3 above)
Comment (onecer) [reply v1.16-4.8/Dockerfile]: bash has been moved to $BUILD_DEPS, after build it will be removed. @OverOrion

File: v1.16-4.8/Dockerfile (hunk 1 above)
Comment (pepov) [reply v1.16-4.8/Dockerfile:42]: we should add make to that build step again and remove it at the end. the goal is to have all the non-essential output plugins added in the last image layer only

Comment (onecer) [reply v1.16-4.8/Dockerfile]: > we should add make to that build step again and remove it at the end. the goal is to have all the non-essential output plugins added in the last image layer only

Done.

I use this to make an image and update the lock file.


Dockerfile-base
```
# TODO every time the base image is changed please make sure BASEIMAGEDEPS is updated (further down in this file)
# https://gallery.ecr.aws/sumologic/kubernetes-fluentd
FROM public.ecr.aws/sumologic/kubernetes-fluentd:1.16.5-sumo-0-alpine as sumo

FROM ruby:3.2.4-alpine3.18 as base

ARG BUILD_DEPS=" \
      make gcc g++ libc-dev \
      wget bzip2 zlib-dev git linux-headers \
      automake autoconf libtool build-base \
      ruby-dev libc6-compat geoip-dev \
      snappy-dev gnupg bash \
      "

RUN addgroup -S -g 101 fluent && adduser -S -G fluent -u 100 fluent \
    # for log storage (maybe shared with host)
    && mkdir -p /fluentd/log \
    # configuration/plugins path (default: copied from .)
    && mkdir -p /fluentd/etc /fluentd/plugins \
    && chown -R fluent /fluentd && chgrp -R fluent /fluentd \
    && mkdir -p /buffers && chown -R fluent /buffers && gem sources --remove https://rubygems.org/ && gem sources -a https://mirrors.aliyun.com/rubygems/

COPY --from=sumo --chown=fluent:fluent /usr/local/bundle /usr/local/bundle

RUN sed -i 's/dl-cdn.alpinelinux.org/mirrors.aliyun.com/g' /etc/apk/repositories && apk add --no-cache \
        ca-certificates \
        ruby ruby-irb ruby-etc \
        tini libmaxminddb geoip \
        snappy \
 && apk add --no-cache $BUILD_DEPS \
 && apk add --no-cache "libssl3>=3.1.4-r1" \
 && apk add --no-cache "libcrypto3>=3.1.4-r1" \
 && echo 'gem: --no-document' >> /etc/gemrc \
 # BASEIMAGEDEPS
 # need to rebuild these gems to have the extensions installed
 && fluent-gem install \
         bigdecimal:1.4.4 \
         oj:3.16.1 \
         google-protobuf:3.21.12 \
         nio4r:2.7.3 \
         snappy:0.3.0 \
 # The only required fluentd dependency is the label router
 # kubeclient install is upgrading the one from the base image
 && fluent-gem install specific_install -v 0.3.8 \
 && fluent-gem specific_install -l https://github.com/ManageIQ/kubeclient.git --ref 054bff2c5e31a555004be2b3c4d32fb9dc5e6a0f \
 && fluent-gem specific_install -l https://github.com/kube-logging/fluent-plugin-label-router.git --ref 6d9d34a7e3fd73a400a31fb35513781802e7b36d \
 && find /usr/local/bundle/gems/ -newer /etc/gemrc -exec chown fluent:fluent {} \; \
 && rm -rf /usr/local/bundle/cache/* && find /usr/local/bundle -name '*.o' -delete

USER root
```

build image

```
docker build -t gem-build . -f Dockerfile-base
```

and then update 

```
docker run --rm -v .:/app -w /app gembuild bundle update
```

**Comment by pepov** [comment]
@onecer I've made some updates to be able to go with the least possible change radius.

**Comment by pepov** [comment]
Also the way I update the gemfile locks is the following: I run the build locally and copy out the updated lock file from the container to be able to compare changes in an architecture independent way:

```
docker build . -t fluentd:v1.16
docker run --rm -ti fluentd:v1.16 cat Gemfile.outputs.lock > outputs/Gemfile.lock.internal
# do the diff
```

**Review by pepov — APPROVED** [review]

**Comment by pepov** [comment]
fyi we have a few ambigous deps now, could be interesting in case we face any issues
```
aws-eventstream (1.3.0, 1.2.0)
aws-partitions (1.896.0, 1.851.0)
aws-sdk-core (3.191.3, 3.186.0)
aws-sdk-sqs (1.67.0, 1.64.0)
aws-sigv4 (1.8.0, 1.6.1)
base64 (0.2.0, default: 0.1.1)
bigdecimal (3.1.6, default: 3.1.3, 1.4.4)
bundler (2.5.15, default: 2.4.19)
concurrent-ruby (1.3.3, 1.2.3, 1.1.10)
domain_name (0.6.20231109, 0.5.20190701)
drb (2.2.1, default: 2.1.1)
faraday (2.7.12, 1.10.3)
faraday-excon (2.1.0, 1.1.0)
faraday-net_http (3.0.2, 1.0.1)
fluent-plugin-record-modifier (2.2.0, 2.1.1)
fluent-plugin-sumologic_output (1.9.0, 1.8.0)
fluentd (1.16.3, 1.16.2)
google-protobuf (3.25.3, 3.21.12)
http (5.1.1, 4.4.1)
kubeclient (4.10.1, 4.9.3)
minitest (5.22.2, 5.16.3)
mutex_m (0.2.0, default: 0.1.2)
net-imap (0.4.2, 0.3.4.1)
net-smtp (0.4.0, 0.3.4)
prometheus-client (4.2.2, 4.2.1)
racc (1.8.0, default: 1.6.2)
timeout (0.4.0, default: 0.3.1)
tzinfo-data (1.2024.1, 1.2023.4, 1.2023.3)
```

**Comment by pepov** [comment]
sorry, I realized, this should only be introduced with logging operator version 4.9, which means we only want to add this plugin to `v1.16-4.9` which is not yet created.

**Comment by onecer** [comment]
> I've created a separate PR to implement the changes in a new image version, I would prefer that one to be merged: #142

Great

**Pull request closed without merging**